File: report/src/main/java/com/vmlens/report/runelementtype/memoryaccesskey/FieldId.java
package com.vmlens.report.runelementtype.memoryaccesskey;

import com.vmlens.report.description.DescriptionContext;
import com.vmlens.report.description.NeedsDescriptionCallback;

public class FieldId implements MemoryAccessKey {

    private final int fieldId;

    public FieldId(int fieldId) {
        this.fieldId = fieldId;
    }

    @Override
    public String asString(DescriptionContext context) {
        return "static " + context.fieldName(fieldId);
    }

    @Override
    public void addToNeedsDescription(NeedsDescriptionCallback callback) {
        callback.needsField(fieldId);
    }

}

File: sync-bug/src/main/scala/com/anarsoft/race/detection/report/element/runelementtype/memoryaccesskey/FieldId.scala
package com.anarsoft.race.detection.report.element.runelementtype.memoryaccesskey

import com.anarsoft.race.detection.report.description.{DescriptionContext, NeedsDescriptionCallback}


class FieldId(val fieldId: Int) extends MemoryAccessKey {
  override def asString(context: DescriptionContext): String = context.fieldName(fieldId)

  override def addToNeedsDescription(callback: NeedsDescriptionCallback): Unit = {
    callback.needsField(fieldId)
  }

  override def objectHashCodeOption: Option[Long] = None


  private def canEqual(other: Any): Boolean = other.isInstanceOf[FieldId]
  
  override def equals(other: Any): Boolean = other match {
    case that: FieldId =>
      that.canEqual(this) &&
        fieldId == that.fieldId
    case _ => false
  }
  
  override def hashCode(): Int = {
    val state = Seq(fieldId)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
